engine-precompiles/src/alt_bn256.rs

```diff
@@ -67,6 +67,15 @@ mod consts {
     pub(super) const ERR_BIG_ENDIAN: ExitError = ExitError::Other(Borrowed("ERR_BIG_ENDIAN"));
 }
 
+#[cfg(feature = "contract")]
+mod type_arith {
+    pub struct Double<const P: usize>;
+    pub trait Is<const S: usize> {}
+    impl Is<128> for Double<64> {}
+    impl Is<64> for Double<32> {}
+    impl Is<32> for Double<16> {}
```

Comment:
Based on the PR description and the previous impl it looks like `S` is allowed to be more than double the size of `P`, but this impl only allows it to be exactly double the size of it.

Is that right? I guess based on the `concat_low_high` function it doesn't really make sense to leave a gap between low and high, but just making sure this is what we want.

Reply:
`contcat_low_high` is a private function, so we only need to enforce constraints based on our own usage. We only use it when `S` is exactly double `P`, so this is the constraint we enforce. If in the future we want to use it for `S` which is greater than `2P` then these compile-time checks will need to made more general (this is still doable).